### ISSUE TITLE
ScaNN evaluation gives better metrics than BruteForce evaluation

### ISSUE THREAD

**Issue by MaiziXiao** [tensorflow/recommenders#264]
Following the documentation (https://www.tensorflow.org/recommenders/examples/efficient_serving#evaluating_the_approximation), I am trying to compare the performance the ScaNN evaluation and BruteForce evaluation.

Code for BruteForce evaluation:
```
item_embeddings = items_data.batch(batch_size).map(model.candidate_model)
# Override the existing streaming candidate source and turn on the compute_metrics
model.task.factorized_metrics = tfrs.metrics.FactorizedTopK(
    candidates=item_embeddings
)
model.compute_metrics = True
# Need to recompile the model for the changes to take effect.
model.compile()
%time results_eval = model.evaluate(cached_test, return_dict=True)
results_eval
```
which gives me result:
`{'factorized_top_k/top_1_categorical_accuracy': 0.0016201249090954661,
 'factorized_top_k/top_5_categorical_accuracy': 0.006863438058644533,
 'factorized_top_k/top_10_categorical_accuracy': 0.01172381266951561,
 'factorized_top_k/top_50_categorical_accuracy': 0.03605514392256737,
 'factorized_top_k/top_100_categorical_accuracy': 0.05449511110782623,
 'loss': 9786.341796875,
 'regularization_loss': 0,
 'total_loss': 9786.341796875}`

Code for ScaNN evaluation:
```
scann = tfrs.layers.factorized_top_k.ScaNN(num_reordering_candidates=1000)
scann.index(item_embeddings)
model.task.factorized_metrics = tfrs.metrics.FactorizedTopK(
    candidates=scann
)
model.compute_metrics = True
model.compile()
%time scann_result = model.evaluate(cached_test, return_dict=True)
scann_result
```
And I get:
`{'factorized_top_k/top_1_categorical_accuracy': 0.004065040498971939,
 'factorized_top_k/top_5_categorical_accuracy': 0.012725344859063625,
 'factorized_top_k/top_10_categorical_accuracy': 0.019765524193644524,
 'factorized_top_k/top_50_categorical_accuracy': 0.05107811838388443,
 'factorized_top_k/top_100_categorical_accuracy': 0.07484976947307587,
 'loss': 9786.341796875,
 'regularization_loss': 0,
 'total_loss': 9786.341796875}`

Is there any good reason that the ANN approach is performing better than a BruteForce approach?

**Comment by maciejkula** [comment]
This is a little unexpected. Could you try using the `BruteForce` layer for your evaluation, for comparison?

**Comment by MaiziXiao** [comment]
@maciejkula After using `BruteForce `layer I get the same result as my first approach showed above. TBH, I didn't really understand the difference between setting the `candidates `  equal an explicit brute_force layer or a set of item embeddings.
Code I use:
```
brute_force = tfrs.layers.factorized_top_k.BruteForce()
brute_force.index(item_embeddings)
model.task.factorized_metrics = tfrs.metrics.FactorizedTopK(
    candidates=brute_force
)
model.compute_metrics = True
model.compile()
%time brute_force_result = model.evaluate(cached_test, return_dict=True)
brute_force_result
```
